templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,871 @@
+
+            # Note: After `train.report`, in the case of remote storage,
+            # the checkpoint directory will be uploaded to the remote storage.
+            train.report(metrics, checkpoint=checkpoint)
+
+            print(
+                "Checkpoint upload time: ",
+                time.perf_counter() - checkpoint_upload_start,
+            )
+            print(
+                "Total checkpointing time: ",
+                time.perf_counter() - checkpoint_save_start,
+            )
+
+        if perplex < config["stop_perplexity"]:
+            stop_perplexity=config["stop_perplexity"]
+            print(f"Perplexity reached {perplex} < {stop_perplexity}. Stopping.")
+            break
+
+        if config["as_test"]:
+            break
+
+
+
+def parse_args():
+
+    parser = argparse.ArgumentParser(description="Simple example of training script.")
+    parser.add_argument(
+        "-mx",
+        "--mixed-precision",
+        type=str,
+        default="bf16",
+        choices=["no", "fp16", "bf16", "fp8"],
+        help="Whether to use mixed precision. Choose"
+        "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
+        "and an Nvidia Ampere GPU.",
+    )
+
+    parser.add_argument(
+        "--batch-size-per-device",
+        "-bs",
+        type=int,
+        default=16,
+        help="Batch size to use per device.",
+    )
+
+    parser.add_argument(
+        "--stop-perplexity",
+        default=0,
+        type=float,
+        help="Target perplexity to reach after which to stop training. Default is 0. "
+        "If 0, training will not stop on perplexity.",
+    )
+
+    parser.add_argument(
+        "--eval-batch-size-per-device",
+        type=int,
+        default=64,
+        help="Batch size to use per device (For evaluation).",
+    )
+
+    parser.add_argument(
+        "--num-devices", "-nd", type=int, default=4, help="Number of devices to use."
+    )
+
+    parser.add_argument(
+        "--grad_accum", type=int, default=1, help="Gradient accumulation steps."
+    )
+    parser.add_argument(
+        "--train_path",
+         type=str,
+         default="./data/train.jsonl",
+         help="Path to training jsonl file"
+    )
+
+    parser.add_argument(
+        "--test_path", 
+        type=str, 
+        default="./data/test.jsonl",
+        help="Path to testing jsonl file"
+    )
+
+    parser.add_argument(
+        "--special_token_path", 
+        type=str, 
+        default="./data/tokens.json",
+        help="Path to token json file"
+    )
+
+    parser.add_argument(
+        "--no-grad-ckpt",
+        action="store_true",
+        help="If passed, will not use gradient checkpointing.",
+    )
+    parser.add_argument(
+        "--output_dir", 
+        type=str, 
+        default="/mnt/local_storage",
+        help="Path to output directory."
+    )
+
+    # parser.add_argument(
+    #     "--model_name", default="meta-llama/Llama-2-7b-chat-hf", type=str
+    # )
```

Comment:
I think we can remove this if it's not needed.

Reply:
Done